examples/jvm-interop/bridge.c

```diff
@@ -0,0 +1,387 @@
+void incref(uint8_t* bytes, uint32_t alignment)
+{
+    ssize_t *refcount_ptr = ((ssize_t *)bytes) - 1;
+    ssize_t refcount = *refcount_ptr;
+
+    if (refcount != REFCOUNT_READONLY) {
+        *refcount_ptr = refcount + 1;
+    }
+}
+
+// Decrement reference count, given a pointer to the first element in a collection.
+// Then call roc_dealloc if nothing is referencing this collection anymore.
+void decref(uint8_t* bytes, uint32_t alignment)
+{
+    if (bytes == NULL) {
+        return;
+    }
+
+    size_t extra_bytes = (sizeof(size_t) >= (size_t)alignment) ? sizeof(size_t) : (size_t)alignment;
+    ssize_t *refcount_ptr = ((ssize_t *)bytes) - 1;
+    ssize_t refcount = *refcount_ptr;
+
+    if (refcount != REFCOUNT_READONLY) {
+        *refcount_ptr = refcount - 1;
+
+        if (refcount == REFCOUNT_ONE) {
+            void *original_allocation = (void *)(refcount_ptr - (extra_bytes - sizeof(size_t)));
+
+            roc_dealloc(original_allocation, alignment);
+        }
+    }
+}
+
+struct RocBytesI32
+{
+    int32_t *bytes;
+    size_t len;
+    size_t capacity;
+};
+
+struct RocBytesI32 init_rocbytes_i32(int32_t *bytes, size_t len)
+{
+    if (len == 0)
+    {
+        struct RocBytesI32 ret = {
+            .len = 0,
+            .bytes = NULL,
+            .capacity = 0,
+        };
+
+        return ret;
+    }
+    else
+    {
+        size_t refcount_size = sizeof(size_t);
+        ssize_t* data = (ssize_t*)roc_alloc(len + refcount_size, alignof(size_t));
+        data[0] = REFCOUNT_ONE;
+        int32_t *new_content = (int32_t *)(data + 1);
+
+        struct RocBytesI32 ret;
+
+        memcpy(new_content, bytes, len * sizeof(int32_t));
+
+        ret.bytes = new_content;
+        ret.len = len;
+        ret.capacity = len;
+
+        return ret;
+    }
+}
+// RocBytesU8 (List U8)
+
+struct RocBytesU8
+{
+    uint8_t *bytes;
+    size_t len;
+    size_t capacity;
+};
+
+struct RocBytesU8 init_rocbytes_u8(uint8_t *bytes, size_t len)
+{
+    if (len == 0)
+    {
+        struct RocBytesU8 ret = {
+            .len = 0,
+            .bytes = NULL,
+            .capacity = 0,
+        };
+
+        return ret;
+    }
+    else
+    {
+
+        size_t refcount_size = sizeof(size_t);
+        ssize_t* data = (ssize_t*)roc_alloc(len + refcount_size, alignof(size_t));
+        data[0] = REFCOUNT_ONE;
+        uint8_t *new_content = (uint8_t *)(data + 1);
+
+        struct RocBytesU8 ret;
+
+        memcpy(new_content, bytes, len * sizeof(uint8_t));
+
+        ret.bytes = new_content;
+        ret.len = len;
+        ret.capacity = len;
+
+        return ret;
+    }
+}
+
+// RocStr
+
+struct RocStr
+{
+    uint8_t *bytes;
+    size_t len;
+    size_t capacity;
+};
+
+struct RocStr init_rocstr(uint8_t *bytes, size_t len)
+{
+    if (len == 0)
+    {
+        struct RocStr ret = {
+            .len = 0,
+            .bytes = NULL,
+            .capacity = 0,
```

Comment:
Capacity of an empty string actually should be set such that the highest bit is a 1 (It believe you have the value set as the constant `MASK`). In other words, it should be equivalent to the most negative value. This is because an empty string is a small string. If fact, you could just delete this case and have it just use the small string case. That would be correct.

Reply:
yea thanks. made it a small string